spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java

```diff
@@ -0,0 +1,162 @@
+ * delete the whole old data file. The changelog table will have two delete-rows(row1 and row2), and
+ * one insert-row(row1). Row1 is a carry-over row.
+ *
+ * <p>The iterator marks the delete-row and insert-row to be the update-rows. For example, these two
+ * rows
+ *
+ * <ul>
+ *   <li>(id=1, data='a', op='DELETE')
+ *   <li>(id=1, data='b', op='INSERT')
+ * </ul>
+ *
+ * will be marked as update-rows:
+ *
+ * <ul>
+ *   <li>(id=1, data='a', op='UPDATE_BEFORE')
+ *   <li>(id=1, data='b', op='UPDATE_AFTER')
+ * </ul>
+ */
+public class ChangelogIterator implements Iterator<Row>, Serializable {
+  private static final String DELETE = ChangelogOperation.DELETE.name();
+  private static final String INSERT = ChangelogOperation.INSERT.name();
+  private static final String UPDATE_BEFORE = ChangelogOperation.UPDATE_BEFORE.name();
+  private static final String UPDATE_AFTER = ChangelogOperation.UPDATE_AFTER.name();
+
+  private final Iterator<Row> rowIterator;
+  private final int changeTypeIndex;
+  private final List<Integer> partitionIdx;
+
+  private Row cachedRow = null;
+
+  public ChangelogIterator(
+      Iterator<Row> rowIterator, int changeTypeIndex, List<Integer> partitionIdx) {
+    this.rowIterator = rowIterator;
+    this.changeTypeIndex = changeTypeIndex;
+    this.partitionIdx = partitionIdx;
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (cachedRow != null) {
+      return true;
+    }
+    return rowIterator.hasNext();
+  }
+
+  @Override
+  public Row next() {
+    // if there is an updated cached row, return it directly
+    if (updated(cachedRow)) {
+      Row row = cachedRow;
+      cachedRow = null;
+      return row;
+    }
+
+    Row currentRow = currentRow();
+
+    if (rowIterator.hasNext()) {
+      GenericRowWithSchema nextRow = (GenericRowWithSchema) rowIterator.next();
+      cachedRow = nextRow;
+
+      if (updateOrCarryoverRecord(currentRow, nextRow)) {
+        Row[] rows = update((GenericRowWithSchema) currentRow, nextRow);
+
+        currentRow = rows[0];
+        cachedRow = rows[1];
+      }
+    }
+
+    return currentRow;
+  }
+
+  private Row[] update(GenericRowWithSchema currentRow, GenericRowWithSchema nextRow) {
+    GenericInternalRow deletedRow = new GenericInternalRow(currentRow.values());
+    GenericInternalRow insertedRow = new GenericInternalRow(nextRow.values());
+
+    if (isCarryoverRecord(deletedRow, insertedRow)) {
+      // set carry-over rows to null for filtering out later
+      return new Row[] {null, null};
+    } else {
+      deletedRow.update(changeTypeIndex, UPDATE_BEFORE);
+      insertedRow.update(changeTypeIndex, UPDATE_AFTER);
+
+      return new Row[] {
+        RowFactory.create(deletedRow.values()), RowFactory.create(insertedRow.values())
+      };
+    }
+  }
+
+  private boolean isCarryoverRecord(GenericInternalRow deletedRow, GenericInternalRow insertedRow) {
+    // set the change_type to the same value
+    deletedRow.update(changeTypeIndex, "");
```

Comment:
Style: I still don't prefer to do a side-effect like this (mutate the variable in the method) just for comparison.  Caller may be taken by surprise.

One way is probably to construct deletedRow and insertedRow with the "" values to begin with.  But I guess Java will be a bit verbose:

```
    new GenericInternalRow(IntStream.range(0, row.values().length)
        .mapToObj(i -> (i == changeTypeIndex) ? "" : row.values()[i])
        .toArray())
```

Another option is to move the update("") back to the base method?

```
GenericInternalRow deletedRow = new GenericInternalRow(currentRow.values());
deletedRow.update(changeTypeIndex, "");
```

At least in this case when you read the update() method, it is more clear what the state is without having to scroll to the helper methods.

Reply:
I would just define an equals method that ignores ChangeTypeIndex